main.py

```diff
@@ -0,0 +1,490 @@
+
+    @staticmethod
+    def renderText(surface, text, location):
+        whitetext = Window.fonts[2].render(text, 1, (255, 255, 255))
+        Window.fonts[0].set_bold(True)
+        blacktext = Window.fonts[2].render(text, 1, (0, 0, 0))
+        Window.fonts[0].set_bold(False)
+
+        textrect = whitetext.get_rect()
+        for i in range(-3, 4):
+            for j in range(-3, 4):
+                textrect.center = [a + b for a, b in zip(location, (i, j))]
+                surface.blit(blacktext, textrect)
+        textrect.center = location
+        surface.blit(whitetext, textrect)
+
+    def __init__(self, thermal_image=None, filename=None):
+        self.exthandler = WindowHandler()
+        if thermal_image is not None:
+            mat = thermal_image.thermal_np.astype(np.float32)
+        
+            if mat.shape != (512, 640):
+                y0, x0 = mat.shape
+                mat = zoom(mat, [512 / y0, 640 / x0])
+
+            self.mat = mat
+            self.mat_orig = mat.copy()
+            self.mat_emm = mat.copy()
+            self.raw = thermal_image.raw_sensor_np
+            self.meta = thermal_image.meta
+            self.overlays = pygame.Surface((640, 512), pygame.SRCALPHA)
+        else:
+            with open(filename, "rb") as f:
+                data = pickle.load(f)
+            self.mat = data.mat
+            self.mat_orig = data.mat_orig
+            self.mat_emm = data.mat_emm
+            self.raw = data.raw
+            self.meta = data.meta
+            self.overlays = pygame.image.fromstring(data.overlays, (640, 512), "RGBA")
+
+            for entry in data.tableEntries:
+                self.exthandler.addToTable(entry)
+            self.exthandler.loadGraph(data.plots)
+
+        self.colorMap = "jet"
+        self.lineNum = 0
+        self.boxNum = 0
+        self.spotNum = 0
+        self.areaMode = "poly"
+        self.selectionComplete = False
+        self.work("colorMap", self.colorMap)
+
+        self.mode = "main"
+        # Dictionary of pages. Each page is a manager.
+        self.managers = {}
+        self.managers["main"] = Manager(
+            buttons=[
+                ((15, 15), (215, 45), "Spot marking", lambda: self.changeMode("spot")),
+                ((15, 75), (215, 45), "Line measurement", lambda: self.changeMode("line")),
+                ((15, 135), (215, 45), "Area marking", lambda: self.changeMode("area")),
+                ((15, 195), (215, 45), "ROI scaling", lambda: self.changeMode("scale")),
+                ((15, 255), (215, 45), "Change colorMap", lambda: self.changeMode("colorMap")),
+                ((15, 315), (215, 45), "Emissivity scaling", lambda: self.changeMode("emissivity")),
+                ((15, 470), (215, 45), "Reset modifications", lambda: self.work("reset")),
+                ((15, 530), (100, 45), "Open image", lambda: self.work("open")),
+                ((130, 530), (100, 45), "Save image", lambda: saveImage(self)),
```

Comment:
Pressing "Save image" and then selecting yes to saving measured values saves 4 files on my PC, one jpg, one pkl, one csv, and one extension-less file. Is the last file required?

Reply:
Additionally, it doesn't save the most recent measurement in the picture. Ex: If the most recent measurement is a line, that line isn't drawn on the output image.